### ISSUE TITLE
Spectral/MT distribution mismatch

### ISSUE THREAD

**Issue by HansLambda** [KlugerLab/bipca#28]
Hi, really exciting work and great initiative. I ran bipca on 200'000 snRNAseq cells and got a somewhat different distribution in the MP histogram. The math behind this is out of my depth but since this method is hyperparameter free I'm wondering if this outcome would be considered ok to continue downstream analysis.

I converted a Seurat object with _sceasy::convertFormat_ to _anndata_, then followed the pbmc tutorial.

[bipca_hist_filtered.pdf](https://github.com/user-attachments/files/19007976/bipca_hist_filtered.pdf)

```
>>> adata.X
<Compressed Sparse Column sparse matrix of dtype 'float64'
        with 479016565 stored elements and shape (203376, 24009)>
>>> X_sub
<Compressed Sparse Column sparse matrix of dtype 'float64'
        with 478947585 stored elements and shape (203376, 22267)>
###BiPCA output
Calculating BiPCA fit...
  Calculating variance fit over 5 submatrices...
    Chebyshev approximation of KS reached 8.023738442809903e-10 with 65 coefficients
    Estimated b=0.0, c=3.573117867323772, KS=0.024274177304754896
    Chebyshev approximation of KS reached 8.494308403602433e-11 with 65 coefficients
    Estimated b=0.05325358788566108, c=3.5913444420296, KS=0.021693978257769198
    Chebyshev approximation of KS reached 5.175458581764947e-11 with 65 coefficients
    Estimated b=0.026331811675289012, c=3.7147133824266287, KS=0.02201173505670917
    Chebyshev approximation of KS reached 1.7509297464512246e-09 with 65 coefficients
    Estimated b=0.0, c=3.6457747298667478, KS=0.02300375275848543
    Chebyshev approximation of KS reached 5.190266338656218e-10 with 65 coefficients
    Estimated b=0.03433066274622635, c=3.5700686483780433, KS=0.023336315222150072
    Approximating the mean of all submatrices
    Approximation ratio is 4.842952243692524e-10 with 65 coefficients
    b=0.03298256071309742, c=3.582021244889728
  Calculated variance fit over 5 submatrices in 622.92 seconds.
  Calculating Sinkhorn biscaling with dense torch backend...
    Sinkhorn converged early after 30 iterations.
  Calculated Sinkhorn biscaling with dense torch backend in 146.10 seconds.
  Calculating dense Biscaling transform...
  Calculated dense Biscaling transform in 7.54 seconds.
  Calculating rank k=22267 dense exact singular value decomposition using __compute_torch_svd....
  Calculated rank k=22267 dense exact singular value decomposition using __compute_torch_svd. in 1382.73 seconds.
  Calculating Shrinker fit...
    Calculating MP Parameter estimate...
      Estimated noise variance computed from the 50th percentile is 1.110
      Scaled Marcenko-Pastur rank is 401
Calculated BiPCA fit in 2190.32 seconds.
Calculating Shrinking singular values according to frobenius loss...
Calculating plotting spectra...
  Calculating spectrum of biwhitened data...
  Calculated spectrum of biwhitened data in 0.02 seconds.
Calculated plotting spectra in 0.02 seconds.

Calculating Writing bipca to anndata...
  Calculating Shrinking singular values according to frobenius loss...
Calculated Writing bipca to anndata in 37.01 seconds.
AnnData object with n_obs × n_vars = 203376 × 22267
    uns: 'bipca'
    obsm: 'U_biwhite'
    varm: 'V_biwhite'
    layers: 'Y_biwhite', 'Z_biwhite'

(<Figure size 3000x1500 with 1 Axes>, <Axes: title={'center': 'Biwhitened covariance $\\frac{{1}}{{N}}YY^T$'}, xlabel='Eigenvalue', ylabel='Density'>)
```

# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             5.1                       1_gnu  
allcools                  1.1.1                    pypi_0    pypi
anndata                   0.11.3                   pypi_0    pypi
annotated-types           0.7.0                    pypi_0    pypi
array-api-compat          1.10.0                   pypi_0    pypi
biomart                   0.9.2                    pypi_0    pypi
bipca                     0.1.dev948+g8b43ef7          pypi_0    pypi
bzip2                     1.0.8                h5eee18b_6  
ca-certificates           2024.12.31           h06a4308_0  
certifi                   2025.1.31                pypi_0    pypi
cffi                      1.17.1                   pypi_0    pypi
charset-normalizer        3.4.1                    pypi_0    pypi
click                     8.1.8                    pypi_0    pypi
cloudpickle               3.1.1                    pypi_0    pypi
contourpy                 1.3.1                    pypi_0    pypi
cycler                    0.12.1                   pypi_0    pypi
dask                      2025.2.0                 pypi_0    pypi
deprecated                1.2.18                   pypi_0    pypi
et-xmlfile                2.0.0                    pypi_0    pypi
exceptiongroup            1.2.2                    pypi_0    pypi
expat                     2.6.4                h6a678d5_0  
filelock                  3.17.0                   pypi_0    pypi
fonttools                 4.56.0                   pypi_0    pypi
fsspec                    2025.2.0                 pypi_0    pypi
h5py                      3.13.0                   pypi_0    pypi
idna                      3.10                     pypi_0    pypi
igraph                    0.11.8                   pypi_0    pypi
imageio                   2.37.0                   pypi_0    pypi
importlib-metadata        8.6.1                    pypi_0    pypi
iniconfig                 2.0.0                    pypi_0    pypi
jinja2                    3.1.5                    pypi_0    pypi
joblib                    1.4.2                    pypi_0    pypi
kiwisolver                1.4.8                    pypi_0    pypi
ld_impl_linux-64          2.40                 h12ee557_0  
legacy-api-wrap           1.4.1                    pypi_0    pypi
leidenalg                 0.10.2                   pypi_0    pypi
libffi                    3.4.4                h6a678d5_1  
libgcc-ng                 11.2.0               h1234567_1  
libgomp                   11.2.0               h1234567_1  
libmpdec                  4.0.0                h5eee18b_0  
libstdcxx-ng              11.2.0               h1234567_1  
libuuid                   1.41.5               h5eee18b_0  
llvmlite                  0.44.0                   pypi_0    pypi
locket                    1.0.0                    pypi_0    pypi
loguru                    0.7.3                    pypi_0    pypi
markupsafe                3.0.2                    pypi_0    pypi
matplotlib                3.10.0                   pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
mudata                    0.3.1                    pypi_0    pypi
muon                      0.1.7                    pypi_0    pypi
mypy-extensions           1.0.0                    pypi_0    pypi
natsort                   8.4.0                    pypi_0    pypi
ncurses                   6.4                  h6a678d5_0  
networkx                  3.4.2                    pypi_0    pypi
numba                     0.61.0                   pypi_0    pypi
numpy                     1.24.4                   pypi_0    pypi
nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
openpyxl                  3.1.5                    pypi_0    pypi
openssl                   3.0.15               h5eee18b_0  
opentsne                  1.0.2                    pypi_0    pypi
packaging                 24.2                     pypi_0    pypi
pandas                    2.2.3                    pypi_0    pypi
pandas-plink              2.3.1                    pypi_0    pypi
pandas-stubs              2.2.3.241126             pypi_0    pypi
pandera                   0.22.1                   pypi_0    pypi
partd                     1.4.2                    pypi_0    pypi
patsy                     1.0.1                    pypi_0    pypi
pillow                    11.1.0                   pypi_0    pypi
pip                       25.0.1                   pypi_0    pypi
pluggy                    1.5.0                    pypi_0    pypi
protobuf                  5.29.3                   pypi_0    pypi
pyalra                    1.4.0                    pypi_0    pypi
pyarrow                   19.0.1                   pypi_0    pypi
pybedtools                0.11.0                   pypi_0    pypi
pybigwig                  0.3.24                   pypi_0    pypi
pychebfun                 1.0.0                    pypi_0    pypi
pycparser                 2.22                     pypi_0    pypi
pydantic                  2.10.6                   pypi_0    pypi
pydantic-core             2.27.2                   pypi_0    pypi
pynndescent               0.5.13                   pypi_0    pypi
pyparsing                 3.2.1                    pypi_0    pypi
pyreadr                   0.5.2                    pypi_0    pypi
pysam                     0.23.0                   pypi_0    pypi
pytest                    8.3.4                    pypi_0    pypi
python                    3.10.16              he870216_1  
python-dateutil           2.9.0.post0              pypi_0    pypi
pytz                      2025.1                   pypi_0    pypi
pyyaml                    6.0.2                    pypi_0    pypi
readline                  8.2                  h5eee18b_0  
requests                  2.32.3                   pypi_0    pypi
scanpy                    1.11.0                   pypi_0    pypi
scikit-learn              1.5.2                    pypi_0    pypi
scipy                     1.15.2                   pypi_0    pypi
seaborn                   0.13.2                   pypi_0    pypi
session-info2             0.1.2                    pypi_0    pypi
setuptools                75.4.0                   pypi_0    pypi
six                       1.17.0                   pypi_0    pypi
sqlite                    3.45.3               h5eee18b_0  
statsmodels               0.14.4                   pypi_0    pypi
sympy                     1.13.1                   pypi_0    pypi
tasklogger                1.2.0                    pypi_0    pypi
texttable                 1.7.0                    pypi_0    pypi
threadpoolctl             3.5.0                    pypi_0    pypi
tk                        8.6.14               h39e8969_0  
tomli                     2.2.1                    pypi_0    pypi
toolz                     1.0.0                    pypi_0    pypi
torch                     2.6.0                    pypi_0    pypi
tqdm                      4.67.1                   pypi_0    pypi
triton                    3.2.0                    pypi_0    pypi
typeguard                 4.4.2                    pypi_0    pypi
types-pytz                2025.1.0.20250204          pypi_0    pypi
typing-extensions         4.12.2                   pypi_0    pypi
typing-inspect            0.9.0                    pypi_0    pypi
tzdata                    2025.1                   pypi_0    pypi
umap-learn                0.5.7                    pypi_0    pypi
urllib3                   2.3.0                    pypi_0    pypi
wheel                     0.45.1          py310h06a4308_0  
wrapt                     1.17.2                   pypi_0    pypi
xarray                    2025.1.2                 pypi_0    pypi
xz                        5.6.4                h5eee18b_1  
zipp                      3.21.0                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_1  
zstandard                 0.22.0                   pypi_0    pypi

**Comment by HansLambda** [comment]
Excellent question, I went back to the count matrix and apparently the conversion tool has pulled the normalized data instead of raw counts. I wasn't diligent enough to check that the first time around.

The results in my initial run were on transformed data. I have re-run the model on raw counts, which looks better:

![Image](https://github.com/user-attachments/assets/90bea46e-5bb2-45b0-bad4-7f5c3a1034f7)

The clustering looks very convincing and runs faster than previous methods where I had to use integrated data.
